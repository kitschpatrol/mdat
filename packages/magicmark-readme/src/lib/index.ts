export {
	type CheckReadmeFileOptions,
	type CheckReadmeFileReport,
	type CheckReadmeStringOptions,
	type CheckReadmeStringReport,
	checkReadmeFile,
	checkReadmeString,
} from './check'

export {
	type ExpandReadmeFileOptions,
	type ExpandReadmeFileReport,
	type ExpandReadmeStringOptions,
	type ExpandReadmeStringReport,
	expandReadmeFile,
	expandReadmeString,
} from './expand'
