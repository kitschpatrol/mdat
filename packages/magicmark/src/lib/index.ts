export {
	type CheckFileReport,
	type CheckOptions,
	type CheckReport,
	checkAst,
	checkFile,
	checkFiles,
	checkString,
} from './check'
export {
	type ExpandAstOptions,
	type ExpandAstReport,
	type ExpandFileOptions,
	type ExpandFileReport,
	type ExpandFilesOptions,
	type ExpandStringOptions,
	type ExpandStringReport,
	expandAst,
	expandFile,
	expandFiles,
	expandString,
} from './expand'
export { default as log } from './log'
export { type Rules, getSoleRule } from './rules'
export { logCheckReport, logExpandFilesReport } from './utilities'
