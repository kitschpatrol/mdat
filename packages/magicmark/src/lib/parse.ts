import { JSDOM } from 'jsdom'
import json5 from 'json5'
import { type Html, type Text } from 'mdast'
import type { JsonObject } from 'type-fest'

export function parseCommentText(
	text: string,
): { args: JsonObject | undefined; keyword: string } | undefined {
	// TODO use JSDOM here as well?
	// Get the keyword with args
	const match = /^\s*<!--\/*[\s#-]*(.+)\s*-->\s*$/.exec(text)?.at(1)?.trim()
	if (match === undefined) return undefined

	// Get the args
	const parts = /^([^\s({]+)[\s()]*({.+})?\)?$/g.exec(match)
	if (parts === null) return undefined

	const keyword = parts.at(1)
	if (keyword === undefined) return undefined

	const rawArgs = parts.at(2)

	// Use json5 so we can be more forgiving about unquoted keys
	const args = rawArgs ? json5.parse<JsonObject>(rawArgs) : undefined

	return { args, keyword }
}

export function splitHtmlIntoMdastNodes(text: string): Array<Html | Text> {
	const frag = JSDOM.fragment(text)

	return [...frag.childNodes].map((node) => {
		switch (node.nodeType) {
			case node.ELEMENT_NODE: {
				// For element nodes, return outerHTML
				return {
					type: 'html',
					value: (node as Element).outerHTML,
				}
			}

			case node.TEXT_NODE: {
				// For text nodes, return textContent
				return {
					type: 'html',
					value: node.textContent ?? '',
				}
			}

			case node.COMMENT_NODE: {
				// For comment nodes, return the comment
				return {
					type: 'html',
					value: `<!--${node.textContent}-->`,
				}
			}

			default: {
				// TODO problem?
				return {
					type: 'html',
					value: '',
				}
			}
		}
	})
}

// Function getOriginalHtml(node: Node): string {
// 	switch (node.nodeType) {
// 		case node.ELEMENT_NODE: {
// 			// For element nodes, return outerHTML
// 			return (node as Element).outerHTML
// 		}

// 		case node.TEXT_NODE: {
// 			// For text nodes, return textContent
// 			return node.textContent ?? ''
// 		}

// 		case node.COMMENT_NODE: {
// 			// For comment nodes, return the comment
// 			return `<!--${node.textContent}-->`
// 		}

// 		default: {
// 			// TODO problem?
// 			return ''
// 		}
// 	}
// }
