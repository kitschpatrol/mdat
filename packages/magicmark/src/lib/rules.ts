import log from '../lib/log'
import { getSoleRecord } from './utilities'
import { type Root } from 'mdast'
import fs from 'node:fs/promises'
import path from 'node:path'
import { fileURLToPath } from 'node:url'
import plur from 'plur'
import { type JsonObject } from 'type-fest'
import type { Merge } from 'type-fest'
import { type SimplifyDeep } from 'type-fest/source/merge-deep'
import { z } from 'zod'

// Basic interface for comment expanders

/**
 * Strict normalized rules used internally.
 * Rules normalized to a form with async content functions and other default metadata
 * Simplifies processing elsewhere, while retaining flexibility for rule authors
 */
export type NormalizedRule = {
	/**
	 * The order in which the rule should be applied during processing
	 * Helpful if a rule depends on the presence of content generated by another rule
	 * If undefined, application order is 0
	 */
	applicationOrder?: number
	content: (options: JsonObject, ast: Root) => Promise<string>
	/**
	 * The expected order of the keyword in the document relative to other expander comments.
	 * Used for validation purposes.
	 * Leave undefined to order skip validation.
	 */
	order?: number
	/**
	 * Whether the presence of the keyword comment in the document is required.
	 * Used for validation purposes.
	 * Leave undefined to skip presence validation.
	 */
	required?: boolean
}

// More flexible rules used in the public interface
export type Rule =
	| Merge<
			NormalizedRule,
			{
				/**
				 * Gets content to expand into the comment.
				 * Can be a simple string for direct replacement, a function that returns a string, or an async function that returns a string.
				 *
				 * If a function is provided, it will be passed the following arguments:
				 *
				 * @param options
				 * JSON object of options parsed from the comment site, e.g.:
				 * `<!-- keyword({something: true}) -->` or
				 * `<!-- keyword {something: true}-->`
				 * Sets options to {something: true}
				 *
				 * @param ast
				 * Markdown abstract syntax tree containing the entire parsed document. Useful for expanders that need the entire document context, such as when generating a table of contents. Do not mutate the AST, instead return a new string.
				 *
				 * @returns A string with the generated content. The string will be parsed as markdown and inserted into the document at the comment's location.
				 */
				content: ((options: JsonObject, ast: Root) => Promise<string> | string) | string
			}
	  >
	/**
	 * The markdown string to expand at the comment site.
	 */
	| string

/**
 * Rules are record objects whose keys match strings inside a Markdown comment, and values explain what should be expanded at the comment site.
 *
 * The record value may be a string, or an object containing additional metadata, possibly with a function to invoke to generate content.
 *
 * @example Most basic rule:
 * { basic: 'content' }
 *
 * @example Rule with metadata:
 * { basic-meta: { required: true, content: 'content'} }
 *
 * @example Rule with dynamic content:
 * { basic-date: { content: () => `${new Date().toISOString()}` } }
 */
export type Rules = SimplifyDeep<Record<string, Rule>>

export type NormalizedRules = SimplifyDeep<Record<string, NormalizedRule>>

/**
 *
 * @param rules - An array that can mix string that are paths to rule .js module files or .json files
 * that will be dynamically loaded with Rules objects that have already been
 * loaded. Rules will be merged from left to right, so later rules will override
 * earlier rules. Also normalizes shorthand rules into a consistent shape.
 * @returns
 */
export async function loadRules(rules: Array<Rules | string>): Promise<NormalizedRules> {
	// Load modules as needed
	let finalRules: Rules = {}

	for (const ruleSetOrRuleModulePath of rules) {
		if (typeof ruleSetOrRuleModulePath === 'string') {
			// Load module or json
			const ruleModulePath = ruleSetOrRuleModulePath
			const fullPath = path.resolve(process.cwd(), ruleModulePath)
			const extension = path.extname(ruleModulePath).toLowerCase()

			if (extension === '.json') {
				// Load json...
				// Every path to a primitive becomes a rule...
				const jsonString = await fs.readFile(fullPath, 'utf8')
				const jsonObject = JSON.parse(jsonString) as JsonObject
				const jsonRules = flattenJson(jsonObject)

				validateRules(jsonRules)

				finalRules = { ...finalRules, ...jsonRules }
			} else if (extension === '.js') {
				// Load module
				const { default: ruleModule } = (await import(
					fileURLToPath(new URL(`file://${fullPath}`))
				)) as { default: Rules }

				// TODO support TS? See SystemJS and SystemJS babel plugin

				const rulesLoadedCount = Object.entries(ruleModule).length

				if (rulesLoadedCount === 0) {
					throw new Error(`No rules found in module: "${ruleModulePath}"`)
				} else {
					validateRules(ruleModule)

					log.info(`Loaded ${rulesLoadedCount} rules from file: "${ruleModulePath}"`)
				}

				finalRules = { ...finalRules, ...ruleModule }
			} else {
				throw new Error(`Unsupported rule file extension: "${extension}"`)
			}
		} else {
			// Merge rule into finalRules
			const ruleSet = ruleSetOrRuleModulePath

			const rulesLoadedCount = Object.entries(ruleSet).length

			if (rulesLoadedCount === 0) {
				throw new Error(`No rules found in internal module`)
			} else {
				validateRules(ruleSet)

				log.info(
					`Loaded ${rulesLoadedCount} ${plur('rule', rulesLoadedCount)} from internal module`,
				)
			}

			finalRules = { ...finalRules, ...ruleSet }
		}
	}

	const finalRulesLoadedCount = Object.entries(finalRules).length
	if (finalRulesLoadedCount === 0) {
		throw new Error(`No rules found. Did you forget to pass in paths to a rules module?`)
	}

	log.info(
		`Loaded ${finalRulesLoadedCount} magicmark comment expansion ${plur('rule', finalRulesLoadedCount)} in total`,
	)

	const normalizedRules = normalizeRules(finalRules)

	return normalizedRules
}

// Helpers
function normalizeRules(rules: Rules): NormalizedRules {
	const normalizedRules: NormalizedRules = {}

	for (const [keyword, rule] of Object.entries(rules)) {
		if (typeof rule === 'string') {
			// Rule is just a simple string replacement
			normalizedRules[keyword] = {
				// eslint-disable-next-line @typescript-eslint/require-await
				content: async () => rule,
			}
		} else if (typeof rule.content === 'string') {
			// Rule is a string replacement with metadata
			// Merge any existing metadata, but turn content into a function
			const ruleContent = rule.content // Needed for type narrowing
			normalizedRules[keyword] = {
				...rule,
				// eslint-disable-next-line @typescript-eslint/require-await
				content: async () => ruleContent,
			}
		} else {
			// Rule content returns a function, wrapped so it can be sync or async
			const ruleContent = rule.content // Needed for type narrowing
			normalizedRules[keyword] = {
				...rule,
				content: async (options: JsonObject, ast: Root) => ruleContent(options, ast),
			}
		}
	}

	return normalizedRules
}

function validateRules(rules: Rules) {
	// Validate, throws on errors
	try {
		normalizedRulesSchema.parse(rules)
	} catch (error) {
		if (error instanceof Error) {
			throw new TypeError(`Error validating rules: ${error.message}`)
		}
	}
}

// Some duplication here, but less painful than inferring the TS types
// from more extensive Zod schemas that would require more extensive duplication
// of the JsonObject and Root types.
const normalizedRulesSchema = z.record(
	z.union([
		z.object({
			applicationOrder: z.number().optional(),
			content: z.union([
				z.string(),
				z
					.function(z.tuple([z.union([z.any(), z.undefined()]), z.union([z.any(), z.undefined()])]))
					.returns(z.union([z.promise(z.string()), z.string()])),
			]),
			order: z.number().optional(),
			required: z.boolean().optional(),
		}),
		z.string(),
	]),
)

// Turns a nested json object into a flat object with dot notation keys
function flattenJson(
	jsonObject: JsonObject,
	parentKey: keyof JsonObject = '',
	result: Record<string, string> = {},
): Record<string, string> {
	for (const [key, value] of Object.entries(jsonObject)) {
		const fullPath = parentKey ? `${parentKey}.${key}` : key

		if (typeof value === 'object' && value !== null && !Array.isArray(value)) {
			flattenJson(value as JsonObject, fullPath, result)
		} else if (value === null) {
			result[fullPath] = 'null'
		} else {
			result[fullPath] = value.toString()
		}
	}

	return result
}

export function getSoleRule<T extends NormalizedRules | Rules>(rules: T): T[keyof T] {
	return getSoleRecord<T[keyof T]>(rules as Record<string, T[keyof T]>)
}
