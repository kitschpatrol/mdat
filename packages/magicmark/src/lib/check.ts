/* eslint-disable max-depth */
/* eslint-disable complexity */
import { parseCommentText } from './parse'
import { type NormalizedRules, type Rules, loadRules } from './rules'
import { expandPath } from './utilities'
import chalk from 'chalk'
import { type Root } from 'mdast'
import fs from 'node:fs/promises'
import plur from 'plur'
import { remark } from 'remark'
import remarkGfm from 'remark-gfm'

export type CheckOptions = {
	meta?: boolean | undefined
	prefix?: string | undefined
	rules: Array<Rules | string>
}
export type CheckReport = Error[] | true
export type CheckFileReport = { file: string; report: CheckReport }

// Check multiple files, main entry point for CLI
export async function checkFiles(
	sourcePaths: string[],
	options: CheckOptions,
): Promise<CheckFileReport[]> {
	const resolvedFiles = sourcePaths.map((path) => expandPath(path))

	if (resolvedFiles.length === 0) {
		throw new Error('No files provided.')
	}

	const results: CheckFileReport[] = []

	for (const file of resolvedFiles) {
		const result = await checkFile(file, options)
		results.push(result)
	}

	return results
}

// Check a file
export async function checkFile(
	sourcePath: string,
	options: CheckOptions,
): Promise<CheckFileReport> {
	const resolvedFile = expandPath(sourcePath)
	const { meta, prefix, rules } = options

	const markdown = await fs.readFile(resolvedFile, 'utf8')
	const result = await checkString(markdown, {
		meta,
		prefix,
		rules,
	})
	return {
		file: resolvedFile,
		report: result,
	}
}

// Check a markdown string
export async function checkString(markdown: string, options: CheckOptions): Promise<CheckReport> {
	const ast = remark().use(remarkGfm).parse(markdown)
	return checkAst(ast, options)
}

// Check an AST
export async function checkAst(ast: Root, options: CheckOptions): Promise<CheckReport> {
	const { meta, prefix = '', rules } = options
	const errors: Error[] = []

	// Load dynamic rule modules if necessary
	const resolvedRules = await loadRules(rules)

	// Get valid expander comments, verify valid args
	const validExpanders: NormalizedRules[] = []
	for (const node of ast.children) {
		if (node.type === 'html') {
			const parsedComment = parseCommentText(node.value)
			if (parsedComment === undefined) continue
			const { args, keyword: prefixedKeyword } = parsedComment

			// Check for prefix
			if (prefix !== '' && !prefixedKeyword.startsWith(prefix)) continue

			// Strip prefix
			const keyword = prefixedKeyword.startsWith(prefix)
				? prefixedKeyword.slice(prefix.length)
				: prefixedKeyword

			// Look for a matching rule in the rule set
			const matchingRule = resolvedRules[keyword]
			if (matchingRule === undefined) continue

			// Comment keyword is in the rule set, check args if present
			// TODO Check if the matching rule can even take args?
			if (args) {
				// See if the rule can run with the args, useful if the rule uses zod or similar
				// to validate the args
				try {
					await matchingRule.content(args, ast)
				} catch (error) {
					if (error instanceof Error) {
						errors.push(error)
					}
				}
			}

			validExpanders.push({ keyword: matchingRule })
		}
	}

	// Check for missing required rules
	const absenceErrors: Error[] = []
	for (const [keyword, rule] of Object.entries(resolvedRules)) {
		if (rule.required && !Object.keys(validExpanders).includes(keyword)) {
			absenceErrors.push(new Error(`  "${chalk.yellow(`<-- ${keyword} -->`)}"`))
		}
	}

	if (absenceErrors.length > 0) {
		errors.push(
			new Error(
				chalk.bold.red(
					`${absenceErrors.length} required ${plur('comment', absenceErrors.length)} ${absenceErrors.length > 1 ? 'are' : 'is'} missing from the document:`,
				),
			),
			...absenceErrors,
		)
	}

	// Check for order issues
	const validExpandersWithOrder = validExpanders.filter((rule) => {
		const theRule = getRuleFromRulesRecord(rule)
		return theRule.order !== undefined
	})

	if (validExpandersWithOrder.length > 1) {
		const sortedValidExpanders = [...validExpandersWithOrder].sort((a, b) => {
			const ruleA = getRuleFromRulesRecord(a)
			const ruleB = getRuleFromRulesRecord(b)
			// Force unwrap because we've checked for it in the filter above
			return ruleA.order! - ruleB.order!
		})

		const sortErrors: Error[] = []

		for (const sortedExpander of validExpandersWithOrder) {
			const correctIndex = sortedValidExpanders.indexOf(sortedExpander)
			const currentIndex = validExpandersWithOrder.indexOf(sortedExpander)
			if (correctIndex !== currentIndex) {
				const up = correctIndex < currentIndex
				let message = `\t${up ? '↑' : '↓'} "${chalk.yellow(sortedExpander.keyword)}" should move ${up ? 'up' : 'down'} so it comes`

				if (correctIndex === 0) {
					message = `${message} before "${chalk.yellow(sortedValidExpanders[correctIndex + 1].keyword)}"`
				} else if (correctIndex === validExpandersWithOrder.length - 1) {
					message = `${message} after "${chalk.yellow(sortedValidExpanders[correctIndex - 1].keyword)}"`
				} else {
					message = `${message} after "${chalk.yellow(sortedValidExpanders[correctIndex - 1].keyword)}" and before "${chalk.yellow(sortedValidExpanders[correctIndex + 1].keyword)}"`
				}

				sortErrors.push(new Error(message))
				continue
			}
		}

		if (sortErrors.length > 0) {
			errors.push(
				new Error(
					chalk.bold.red(
						`${sortErrors.length} ${plur('comment', sortErrors.length)} ${sortErrors.length > 1 ? 'are' : 'is'} not in the correct order:`,
					),
				),
				...sortErrors,
			)
		}

		if (meta) {
			// TODO check for meta comment
		}

		// TODO maybe check for missing closing tags, indicating that the tool hasn't been run?
	}

	if (errors.length > 0) {
		return errors
	}

	return true
}

// Helper for working with arrays of rule records that should only have a single rule.
function getRuleFromRulesRecord(rulesRecordArray: NormalizedRules): NormalizedRules[keyof Rules] {
	const rules = Object.values(rulesRecordArray)
	if (rules.length > 1) {
		throw new Error(
			'Found multiple rule definitions in a valid rule array entry. This should never happen',
		)
	}

	return rules[0]
}
