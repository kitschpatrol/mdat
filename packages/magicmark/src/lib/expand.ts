import { parseCommentText } from './parse'
import { type Rules, loadRules } from './rules'
import { getInputOutputPath, getInputOutputPaths } from './utilities'
import { type Html, type Root } from 'mdast'
import fs from 'node:fs/promises'
import { remark } from 'remark'
import remarkGfm from 'remark-gfm'
import type { JsonObject } from 'type-fest'
import { type Simplify } from 'type-fest'
import { CONTINUE, EXIT, visit } from 'unist-util-visit'

export type ExpandFileReport = { expandedFile: string; report: string[]; sourceFile: string }
export type ExpandStringReport = { expandedString: string; report: string[] }
export type ExpandAstReport = { expandedAst: Root; report: string[] }

// Multiple files, main entry point for CLI
export type ExpandFilesOptions = ExpandFileOptions
export async function expandFiles(
	sourcePaths: string[],
	options: ExpandFilesOptions,
): Promise<ExpandFileReport[]> {
	const { meta, name, output, prefix, print, rules } = options
	const inputOutputPaths = getInputOutputPaths(sourcePaths, output, name)
	const results: ExpandFileReport[] = []

	for (const { input, output } of inputOutputPaths) {
		const result = await expandFile(input, {
			meta,
			output,
			prefix,
			print,
			rules,
		})
		results.push(result)
	}

	return results
}

// Single file
export type ExpandFileOptions = Simplify<
	ExpandStringOptions & {
		name?: string
		output?: string
		print?: boolean
	}
>
export async function expandFile(
	sourcePath: string,
	options: ExpandFileOptions,
): Promise<ExpandFileReport> {
	const { meta, name, output, prefix, print, rules } = options
	const { input: inputPath, output: outputPath } = getInputOutputPath(sourcePath, output, name)

	const markdown = await fs.readFile(inputPath, 'utf8')
	const { expandedString, report } = await expandString(markdown, {
		meta,
		prefix,
		rules,
	})

	if (print) {
		process.stdout.write(`${expandedString}\n`)
	} else {
		await fs.writeFile(outputPath, expandedString)
	}

	return {
		expandedFile: print ? 'stdout' : outputPath,
		report,
		sourceFile: inputPath,
	}
}

// String
export type ExpandStringOptions = ExpandAstOptions
export async function expandString(
	markdown: string,
	options: ExpandStringOptions,
): Promise<ExpandStringReport> {
	const ast = remark().use(remarkGfm).parse(markdown)
	const { expandedAst, report } = await expandAst(ast, options)

	return {
		expandedString: remark().use(remarkGfm).stringify(expandedAst),
		report,
	}
}

// AST
export type ExpandAstOptions = {
	meta?: boolean
	prefix?: string
	rules: Array<Rules | string>
}
export async function expandAst(ast: Root, options: ExpandAstOptions): Promise<ExpandAstReport> {
	const { meta = false, prefix = '', rules } = options

	// Load dynamic rule modules if necessary, and also normalizes the rule type
	const resolvedRules = await loadRules(rules)

	// Extract template expansion commands from comment nodes
	// https://github.com/syntax-tree/mdast/blob/main/readme.md

	// Happens in two passes to accommodate async rules

	// Save promises as we go
	const newContent: Array<{
		applySequence: number
		args: JsonObject | undefined
		// Convert all rule content to this form
		content: (options: JsonObject, ast: Root) => Promise<string>
		openingComment: Html
	}> = []

	// TODO stay at top level
	// First pass adds closing comments if needed and removes existing children
	visit(ast, 'html', (node, index, parent) => {
		if (parent !== undefined && index !== undefined) {
			// Parse the comment contents
			const result = parseCommentText(node.value)
			if (result === undefined) return CONTINUE
			const { args, keyword: prefixedKeyword } = result

			// Check for prefix
			if (prefix !== '' && !prefixedKeyword.startsWith(prefix)) return CONTINUE

			// Strip prefix
			const keyword = prefixedKeyword.startsWith(prefix)
				? prefixedKeyword.slice(prefix.length)
				: prefixedKeyword

			// Look for a matching rule in the rule set
			const matchingRule = resolvedRules[keyword]
			if (matchingRule === undefined) return CONTINUE

			// Look for a closing closing comment, if there is one we'll delete everything
			// between the opening and closing comments
			const closingTagIndex = getClosingTagIndex(ast, index, prefixedKeyword)
			const tagsToReplace = closingTagIndex ? closingTagIndex - index - 1 : 0
			parent.children.splice(index + 1, tagsToReplace)

			// Add closing tag if it doesn't exist
			if (closingTagIndex === undefined) {
				const closingNode: Html = {
					type: 'html',
					value: `<!-- /${prefixedKeyword} -->`,
				}
				parent.children.splice(index + 1, 0, closingNode)
			}

			newContent.push({
				applySequence: matchingRule.applicationOrder ?? 0,
				args,
				content: matchingRule.content,
				openingComment: node,
			})
		}
	})

	// Sort newContent in place to apply expansion rules in the order they're received in the array
	newContent.sort((a, b) => a.applySequence - b.applySequence)

	// Execution, not just promise resolution, must be deferred to here
	// to ensure table of contents has all generated headings
	for (const { args, content, openingComment } of newContent) {
		const newMarkdownString = await content(args ?? {}, ast)
		const newNodes = remark().use(remarkGfm).parse(newMarkdownString).children
		const openingCommentIndex = ast.children.indexOf(openingComment)
		ast.children.splice(openingCommentIndex + 1, 0, ...newNodes)
	}

	// Add or remove the meta comment if requested
	const message = 'Warning: Content in HTML comment blocks generated by magicmark'
	// Remove existing meta comments
	ast.children = ast.children.filter(
		(node) => !(node.type === 'html' && node.value.startsWith(`<!-- ${message}`)),
	)

	if (meta) {
		const date = new Date().toISOString().slice(0, 10)
		const metaComment: Html = {
			type: 'html',
			value: `<!-- ${message} on ${date}. -->`,
		}
		ast.children.unshift(metaComment)
	}

	return {
		expandedAst: ast,
		report: newContent.map(({ openingComment }) => openingComment.value),
	}
}

// Helpers
function getClosingTagIndex(
	ast: Root,
	startFromIndex: number,
	prefixedKeyword: string,
): number | undefined {
	let matchingIndex: number | undefined
	visit(ast, 'html', (node, index, parent) => {
		if (parent !== undefined && index !== undefined && index >= startFromIndex) {
			const result = parseCommentText(node.value)
			if (result === undefined) return CONTINUE

			if (`/${prefixedKeyword}` === result.keyword) {
				matchingIndex = index
				return EXIT
			}
		}
	})
	return matchingIndex
}
