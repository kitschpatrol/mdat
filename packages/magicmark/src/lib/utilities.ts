import { type CheckFileReport } from './check'
import { type ExpandFileReport } from './expand'
import log from './log'
import logSymbols from 'log-symbols'
import fs from 'node:fs'
import path from 'node:path'
import { isFileSync } from 'path-type'
import plur from 'plur'
import untildify from 'untildify'

export function getInputOutputPaths(
	inputs: string[],
	output: string | undefined,
	name: string | undefined,
	extension: string | undefined,
): Array<{ input: string; name: string; output: string }> {
	const paths: Array<{ input: string; name: string; output: string }> = []

	// Accounts for numbering outputs if multiple files are provided
	// TODO zero pad if more than 9 files
	for (const [index, file] of inputs.entries()) {
		const nameSuffix = name && inputs.length > 1 ? `-${index + 1}` : ''
		const inputOutputPath = getInputOutputPath(file, output, name, extension, nameSuffix)
		paths.push(inputOutputPath)
	}

	return paths
}

export function getInputOutputPath(
	input: string,
	output: string | undefined,
	name: string | undefined,
	extension: string | undefined,
	nameSuffix = '',
): { input: string; name: string; output: string } {
	const resolvedInput = expandPath(input)
	const resolvedOutput = output ? expandPath(output) : undefined

	// Ensure input is a file
	if (!isFileSync(resolvedInput)) {
		throw new Error(`Input file not found: "${resolvedInput}"`)
	}

	console.log(`inner output: ${output}`)

	// Ensure output is not a file
	if (resolvedOutput) {
		if (isFileSync(resolvedOutput)) {
			throw new Error(`Output path must be a directory, received a file path: "${resolvedOutput}"`)
		}

		// Create intermediate directories if needed
		fs.mkdirSync(resolvedOutput, { recursive: true })
	}

	// Get base fileName either from input or name option
	const baseName = name
		? path.basename(name, path.extname(name))
		: path.basename(resolvedInput, path.extname(resolvedInput))

	console.log('----------------------------------')
	console.log(`output: ${output}`)
	console.log(`resolvedOutput: ${resolvedOutput}`)
	console.log(`baseName: ${baseName}`)

	// Use argument first, then output name extension if present, then input name extension if present, then default to nothing
	const resolvedExtension = `.${
		extension ??
		(name && path.extname(name) !== ''
			? path.extname(name)
			: path.extname(input) === ''
				? ''
				: path.extname(input))
	}`

	const outputName = `${baseName}${nameSuffix}${resolvedExtension}`
	const outputPath = resolvedOutput ?? path.dirname(resolvedInput)

	return { input: resolvedInput, name: outputName, output: outputPath }
}

export function expandPath(file: string): string {
	// TODO any other validation?
	return untildify(file)
}

// The log.ts module is intended for more generic uses, so the log formatting
// functions below are going in this file

/*
 * @param report - The report to log, returned from expandFiles
 * @returns The number of expansions made across all files
 */
export function logExpandFilesReport(report: ExpandFileReport[]): number {
	let expansionCount = 0

	for (const { expandedFile, report: fileReport, sourceFile } of report) {
		log.infoPrefixed('expand', 'Expanded:')
		log.infoPrefixed('expand', `  From: ${sourceFile}`)
		log.infoPrefixed('expand', `  To:   ${expandedFile}`)

		const expansionCountInFile = fileReport.length
		expansionCount += expansionCountInFile

		log.infoPrefixed('expand', `  Expansions: ${expansionCountInFile}`)
		for (const [i, line] of fileReport.entries()) {
			log.infoPrefixed('expand', `    ${i + 1}. ${line}`)
		}
	}

	return expansionCount
}

/*
 *
 * @param report - The report to log, returned from checkFiles
 * @returns The number of errors logged
 */
export function logCheckReport(report: CheckFileReport[]): number {
	let errorCount = 0
	for (const { file, report: fileReport } of report) {
		if (typeof fileReport === 'boolean' && fileReport) {
			log.infoPrefixed('check', `${file}: ${logSymbols.success} OK`)
		} else {
			log.errorPrefixed(
				'check',
				`${file}: ${logSymbols.warning} ${fileReport.length} ${plur('Error', fileReport.length)}:`,
			)
			for (const error of fileReport) {
				log.errorPrefixed('check', `   ${logSymbols.error} ${error.message}`)
				errorCount++
			}
		}
	}

	return errorCount
}
