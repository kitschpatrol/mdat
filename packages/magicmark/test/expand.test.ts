import { expandString } from '../src/lib/expand'
import fs from 'node:fs/promises'
import { describe, expect, it } from 'vitest'

describe('comment expansion', () => {
	it('should expand basic comments', async () => {
		const markdown = await fs.readFile('./test/assets/test-document.md', 'utf8')
		const { expandedString } = await expandString(markdown, {
			rules: ['./test/assets/test-rules.js'],
		})

		expect(expandedString).toMatchSnapshot()
	})

	it('should expand prefixed comments only', async () => {
		const markdown = await fs.readFile('./test/assets/test-document.md', 'utf8')
		const { expandedString } = await expandString(markdown, {
			prefix: 'mm-',
			rules: ['./test/assets/test-rules.js'],
		})

		expect(expandedString).toMatchSnapshot()
	})

	it('should include the meta tag if asked', async () => {
		const markdown = await fs.readFile('./test/assets/test-document.md', 'utf8')
		const { expandedString } = await expandString(markdown, {
			meta: true,
			rules: ['./test/assets/test-rules.js'],
		})

		expect(expandedString).toMatchSnapshot()
	})

	it('should throw an error if rule set is invalid', async () => {
		const markdown = await fs.readFile('./test/assets/test-document.md', 'utf8')

		await expect(
			expandString(markdown, {
				meta: true,
				rules: ['./test/assets/test-rules-invalid.js'],
			}),
		).rejects.toThrow()
	})
})
