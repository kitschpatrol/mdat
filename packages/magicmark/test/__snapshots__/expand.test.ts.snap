// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`comment expansion > should expand basic comments 1`] = `
"# The MagicMark sample document

## Basic comment expansion

<!-- basic -->

**A bold statement from test-rules.js**

<!-- /basic -->

## Prefixed comment

Expansion only happens when \`--prefix mm\` option is set:

<!-- mm-basic -->

## Comment expansion with dynamic content

Shows the name of the rules file:

<!-- basic-dynamic -->

I was generated from test-rules.js

<!-- /basic-dynamic -->

## Required expansion

Inclusion is validated via \`--check\` option

<!-- basic-list-required -->

* I
* am
* a
* list
* that
* must
* be
* here

<!-- /basic-list-required -->

## Expansions with options argument, canonical syntax

<!-- basic-options {prefix: "ðŸª´ ", suffix: " ðŸª´"} -->

ðŸª´ I am between two ferns ðŸª´

<!-- /basic-options -->

## Expansions with arguments, syntax forgiveness

<!--- # basic-options({prefix: "ðŸª´ ", suffix: " ðŸª´"}) -->

ðŸª´ I am between two ferns ðŸª´

<!-- /basic-options -->

## Ordered expansion

Order in the document is validated via \`--check\` option

<!-- basic-ordered -->

I had to be last

<!-- /basic-ordered -->
"
`;

exports[`comment expansion > should expand prefixed comments only 1`] = `
"# The MagicMark sample document

## Basic comment expansion

<!-- basic -->

**A bold statement from test-rules.js**

<!-- /basic -->

## Prefixed comment

Expansion only happens when \`--prefix mm\` option is set:

<!-- mm-basic -->

**A bold statement from test-rules.js**

<!-- /mm-basic -->

## Comment expansion with dynamic content

Shows the name of the rules file:

<!-- basic-dynamic -->

I was generated from test-rules.js

<!-- /basic-dynamic -->

## Required expansion

Inclusion is validated via \`--check\` option

<!-- basic-list-required -->

* I
* am
* a
* list
* that
* must
* be
* here

<!-- /basic-list-required -->

## Expansions with options argument, canonical syntax

<!-- basic-options {prefix: "ðŸª´ ", suffix: " ðŸª´"} -->

ðŸª´ I am between two ferns ðŸª´

<!-- /basic-options -->

## Expansions with arguments, syntax forgiveness

<!--- # basic-options({prefix: "ðŸª´ ", suffix: " ðŸª´"}) -->

ðŸª´ I am between two ferns ðŸª´

<!-- /basic-options -->

## Ordered expansion

Order in the document is validated via \`--check\` option

<!-- basic-ordered -->

I had to be last

<!-- /basic-ordered -->
"
`;

exports[`comment expansion > should include the meta tag if asked 1`] = `
"<!-- Warning: Content in HTML comment blocks generated by magicmark on 2024-01-27. -->

# The MagicMark sample document

## Basic comment expansion

<!-- basic -->

**A bold statement from test-rules.js**

<!-- /basic -->

## Prefixed comment

Expansion only happens when \`--prefix mm\` option is set:

<!-- mm-basic -->

## Comment expansion with dynamic content

Shows the name of the rules file:

<!-- basic-dynamic -->

I was generated from test-rules.js

<!-- /basic-dynamic -->

## Required expansion

Inclusion is validated via \`--check\` option

<!-- basic-list-required -->

* I
* am
* a
* list
* that
* must
* be
* here

<!-- /basic-list-required -->

## Expansions with options argument, canonical syntax

<!-- basic-options {prefix: "ðŸª´ ", suffix: " ðŸª´"} -->

ðŸª´ I am between two ferns ðŸª´

<!-- /basic-options -->

## Expansions with arguments, syntax forgiveness

<!--- # basic-options({prefix: "ðŸª´ ", suffix: " ðŸª´"}) -->

ðŸª´ I am between two ferns ðŸª´

<!-- /basic-options -->

## Ordered expansion

Order in the document is validated via \`--check\` option

<!-- basic-ordered -->

I had to be last

<!-- /basic-ordered -->
"
`;
